[llvm-debuginfo-analyzer] Common handling of unsigned attribute values.

CarlosAlbertoEnciso opened the pull request:
- In the DWARF reader, for those attributes that can have an unsigned value, allow for the following cases:
  * Is an implicit constant
  * Is an optional value
- The testing is done by creating a file with generated DWARF, using `DwarfGenerator` (generate DWARF debug info for unit tests).

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Common handling of unsigned attribute values.

- In the DWARF reader, for those attributes that can have an
  unsigned value, allow for the following cases:
  * Is an implicit constant
  * Is an optional value

pogo59 reviewed (commented):
The test makes it pretty clear this is functionally the correct thing to do. But it covers only the inline attribute; are we missing tests for the other affected attributes? (I can't remember whether there is a unittest for the reader, but if there is, that would seem like a good place to cover all these cases.)

CarlosAlbertoEnciso commented:
We have unittest for the reader: `llvm/unittests/DebugInfo/LogicalView`. But we don't have the ability to create logical elements with specific DWARF tags/attributes.
I will look at adding some generic basic support using `llvm/unittests/DebugInfo/DWARF`

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Common handling of unsigned attribute values.

- In the DWARF reader, for those attributes that can have an
  unsigned value, allow for the following cases:
  * Is an implicit constant
  * Is an optional value
- The testing is done by creating a file with generated DWARF,
  using 'DwarfGenerator' (generate DWARF debug info for unit tests).

CarlosAlbertoEnciso commented:
@pogo59 Using the existing `llvm/unittests/DebugInfo/DWARF/DwarfGenerator` module, added a unittest to include the affected attributes.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Common handling of unsigned attribute values.

- In the DWARF reader, for those attributes that can have an
  unsigned value, allow for the following cases:
  * Is an implicit constant
  * Is an optional value
- The testing is done by creating a file with generated DWARF,
  using 'DwarfGenerator' (generate DWARF debug info for unit tests).
- Remove not needed include header file.

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+
+// Helper function to get the first compile unit.
+LVScopeCompileUnit *getFirstCompileUnit(LVScopeRoot *Root) {
+  EXPECT_NE(Root, nullptr);
```
```suggestion
  ASSERT_NE(Root, nullptr);
```
Use the `ASSERT_*` macros when the test will later crash if the condition is not true. This is especially the case when verifying a pointer is non-null and will be dereferenced.

CarlosAlbertoEnciso replied:
Changed all `EXPECT_NE` to `ASSERT_NE`.

**Note**: There is a restriction on using `EXPECT_NE` inside a non-void function
https://chromium.googlesource.com/external/github.com/google/googletest/+/HEAD/docs/advanced.md#assertion-placement

For the non-void functions: `getFirstCompileUnit` and `createReader`, I used standard if checks and the caller will use `ASSERT_NE` on the returned value.

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+// Check the logical elements basic properties.
+void checkElementAttributes(LVReader *Reader) {
+  LVScopeRoot *Root = Reader->getScopesRoot();
+  EXPECT_NE(Root, nullptr);
```
This is redundant with a check inside getFirstCompileUnit

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+LVScopeCompileUnit *getFirstCompileUnit(LVScopeRoot *Root) {
+  EXPECT_NE(Root, nullptr);
+  const LVScopes *CompileUnits = Root->getScopes();
+  EXPECT_NE(CompileUnits, nullptr);
```
```suggestion
  ASSERT_NE(CompileUnits, nullptr);
```

CarlosAlbertoEnciso replied:
The check has been removed from getFirstCompileUnit. See previous comment about assertion-placement.

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+  LVScopes::const_iterator ScopeIter = Scopes->begin();
+  EXPECT_NE(ScopeIter, nullptr);
+  LVScope *Scope = static_cast<LVScope *>(*ScopeIter);
+  EXPECT_NE(Scope, nullptr);
```
```suggestion
  ASSERT_NE(Scope, nullptr);
```

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+  EXPECT_EQ(Scope->getDiscriminator(), 8);     // ScopeFunctionInlined
+
+  // Check no-values.
+  EXPECT_NE(++ScopeIter, nullptr);
```
Another iteration check

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+  // Check no-values.
+  EXPECT_NE(++ScopeIter, nullptr);
+  Scope = static_cast<LVScope *>(*ScopeIter);
+  EXPECT_NE(Scope, nullptr);
```
```suggestion
  ASSERT_NE(Scope, nullptr);
```

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+  // Check implicit values.
+  EXPECT_NE(++ScopeIter, nullptr);
+  Scope = static_cast<LVScope *>(*ScopeIter);
+  EXPECT_NE(Scope, nullptr);
```
```suggestion
  ASSERT_NE(Scope, nullptr);
```

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+  EXPECT_EQ(CompileUnits->size(), 1u);
+
+  LVScopes::const_iterator Iter = CompileUnits->begin();
+  EXPECT_NE(Iter, nullptr);
```
Iterators generally should be compared to `end()` ? Also this should be an `ASSERT_NE`.

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+
+  // Check values.
+  LVScopes::const_iterator ScopeIter = Scopes->begin();
+  EXPECT_NE(ScopeIter, nullptr);
```
Iteration should be compared to `end()` ? Also, use `ASSERT_NE` here.

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+  EXPECT_EQ(Scope->getDiscriminator(), 0);     // ScopeFunctionInlined
+
+  // Check implicit values.
+  EXPECT_NE(++ScopeIter, nullptr);
```
Another iterator check

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+
+  // Check values.
+  LVSymbols::const_iterator SymbolIter = Symbols->begin();
+  EXPECT_NE(SymbolIter, nullptr);
```
```suggestion
  ASSERT_NE(SymbolIter, nullptr);
```

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+  LVSymbols::const_iterator SymbolIter = Symbols->begin();
+  EXPECT_NE(SymbolIter, nullptr);
+  LVSymbol *Symbol = static_cast<LVSymbol *>(*SymbolIter);
+  EXPECT_NE(Symbol, nullptr);
```
```suggestion
  ASSERT_NE(Symbol, nullptr);
```

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:
```diff
@@ -0,0 +1,369 @@
+  EXPECT_EQ(Symbol->getBitSize(), 1); // Symbol
+
+  // Check no-values.
+  EXPECT_NE(++SymbolIter, nullptr);
```
This pattern continues, I won't add comments on all of them.

CarlosAlbertoEnciso replied:
Removed the repeated comments.

pogo59 commented on llvm/unittests/DebugInfo/LogicalView/CMakeLists.txt:
```diff
@@ -1,17 +1,22 @@
 
 add_llvm_unittest_with_input_files(DebugInfoLogicalViewTests
+  ../DWARF/DwarfGenerator.cpp
+  ../DWARF/DwarfUtils.cpp
```
This is okay for now, but really these components should be moved into their own library so they can be shared between unittests.

CarlosAlbertoEnciso replied:
That is a good point. I will create a PR to record that request.

CarlosAlbertoEnciso committed:
[llvm-debuginfo-analyzer] Common handling of unsigned attribute values.

- In the DWARF reader, for those attributes that can have an
  unsigned value, allow for the following cases:
  * Is an implicit constant
  * Is an optional value
- The testing is done by creating a file with generated DWARF,
  using 'DwarfGenerator' (generate DWARF debug info for unit tests).

- Address reviewer comments:
  * Changed 'EXPECT_NE' to 'ASSERT_NE'
  * Test iterators against 'end()'
  * Remove redundant comments
  * As the 'ASSERT_NE' can not be used inside a non-void function,
    use normal logic and move the checks outside the function.
    The caller will use 'ASSERT_NE' on the returned value.

pogo59 reviewed (approved):
LGTM

CarlosAlbertoEnciso commented:
@pogo59 Thanks very much for your valuable review.

CarlosAlbertoEnciso commented:
Looking at the reported failures on those specific builders:
- Link errors: Missing explicitly references to `DebugInfoDWARF` and `Object` in the `CMakeList.txt`
- Compile errors: error: comparison of integers of different signs:

nico commented on llvm/unittests/DebugInfo/LogicalView/CMakeLists.txt:
```diff
@@ -1,17 +1,22 @@
 set(LLVM_LINK_COMPONENTS
+  ${LLVM_TARGETS_TO_BUILD}
```
Doesn't this make the next 3 lines superfluous?

CarlosAlbertoEnciso replied:
@nico That is a good point. With your suggested change, all pre-commit test pass.
Opened https://github.com/llvm/llvm-project/pull/118052
